### PR TITLE
codecs resiliency, another approach

### PR DESCRIPTION
**Pull request by tmszdmsk**
After some thinking about #359 I prefer it this way because:
* it is more composable than writing the same logic in every codec
* covers all built-in and 3rd party codecs

Questions:
* can we assume that this should be the behavior for all possible `Injectors` and `Extractors`? Can you imagine any exception to this rule?
* it would be great to hide it from the external world, i.e. make it at least package scoped. IMHO they "belong" to `propagation` package naturally but if we want to make it "invisible" we can:
  * place them the main package where the `Tracer` with package scope
  * place it inside `Tracer`, but it is quite large already.
  * extract `PropagationRegistry` to a separate file and move decorators inside of it keeping it in the main package with package visibility. *I prefer this one*

I am new to this project so please think if I've missed something important from the design perspective.
 
To do
------
- [ ] tests

### CODE REVIEW

File: jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.uber.jaeger.propagation;
+
+import com.uber.jaeger.SpanContext;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@RequiredArgsConstructor
```
Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java]: Could you add a javadoc, stating why this exists and how it works?

**Comment by jpkrohling** [comment]
The build failed with `License violations were found: /.../jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java,/home/travis/build/jaegertracing/jaeger-client-java/jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingInjectorDecorator.java`

If you want to catch such errors faster, you may want to run `./gradlew check` locally before pushing updates to the PR :)

File: jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java (hunk 1 above)
Comment (tmszdmsk) [reply jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java]: you mean https://projectlombok.org/features/constructor?

**Comment by tmszdmsk** [comment]
@jpkrohling please, wait with low-level CR. I want to establish if it's an accepted way to go on a "design level" first.

File: jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java (hunk 1 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java]: No, sorry, I meant a class-level javadoc for the classes `ExceptionCatching*Decorator`

Comment (tmszdmsk) [reply jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java]: sure, I will do that although I think that  `Injector` and `Extractor` interfaces need more attention in this area.

**Comment by tmszdmsk** [comment]
ok, I've extracted PropagationRegistry to separate class to hide exception-catching decorators from users. 

also, I've added javadoc proposition for `Extractor` which is based on my understanding of this topic. Please comment if I made wrong assumptions.

File: jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+class PropagationRegistry {
+  private final Map<Format<?>, Injector<?>> injectors = new HashMap<Format<?>, Injector<?>>();
+  private final Map<Format<?>, Extractor<?>> extractors = new HashMap<Format<?>, Extractor<?>>();
+
+  @SuppressWarnings("unchecked")
+  <T> Injector<T> getInjector(Format<T> format) {
+    return new ExceptionCatchingInjectorDecorator<T>((Injector<T>) injectors.get(format));
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java]: there's no reason to wrap the injector on every `get`, instead wrap it on `register`

File: jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    injectors.put(format, injector);
+  }
+
+  public <T> void register(Format<T> format, Extractor<T> extractor) {
+    extractors.put(format, extractor);
+  }
+
+  @RequiredArgsConstructor
+  @Slf4j
+  public static class ExceptionCatchingExtractorDecorator<T> implements Extractor<T> {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java]: why public?

File: jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      } catch (RuntimeException ex) {
+        ExceptionCatchingExtractorDecorator.log.warn("Error when extracting SpanContext from carrier. Handling gracefully.", ex);
+        return null;
+      }
+    }
+  }
+
+  @RequiredArgsConstructor
+  @Slf4j
+  public static class ExceptionCatchingInjectorDecorator<T> implements Injector<T> {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java]: why public?

File: jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java (hunk 3 above)
Comment (tmszdmsk) [reply jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java]: yep, automatic refactor ;)

**Comment by yurishkuro** [comment]
Thanks @tmszdmsk !

**Comment by tmszdmsk** [comment]
no problem! ;)

**Pull request merged**